### DISCUSSION TITLE
`assert is not a function` error, when import `csrf` in `middleware.js`

### DISCUSSION THREAD

**Discussion by rwx222** [vercel/next.js#68168] [Help] [answered]
### Summary

Hi everyone 👋,
I'm having the following problem and I was wondering if you could help or give me an idea of what's going on.

You can try this on a fresh Next.js installation. [code example repo](https://github.com/andresdotsh/csrf-next-issue/tree/main) 👈

When I import the `csrf` library in the `middleware.js` file, I get an error that says `assert is not a function`.

This is because the [csrf](https://www.npmjs.com/package/csrf) library depends on the [rndm](https://www.npmjs.com/package/rndm) library, which uses **NodeJS**'s `assert` in its code.

This error **only occurs** if I import it **in the middleware.js file, it doesn't happen in page.js** for example.

Does anyone have any idea why this might be happening? I would like to use this `csrf` library, but this error is preventing me from doing so. Thanks in advance for any help 🙏.

More info:

- `nodejs v20.15.1`
- `next@14.2.5`
- `csrf@3.1.0`
- `yarn@1.22.22`
- I'm using nextjs `app router`

Pictures:

![Error example](https://i.ibb.co/dgZF8kf/err1.png)

![Error example 2](https://i.ibb.co/JqXPsNN/err2.png)

### Example

https://github.com/andresdotsh/csrf-next-issue/tree/main

**Comment by Juneezee** [comment, marked as answer]
That is because Next.js's middleware runs on Edge Runtime [^1]. The `assert` module is only available in Node runtime. See the **Unsupported APIs** section [^2]. Calling `require` in Edge Runtime is not allowed.

You can try the [@edge-csrf/nextjs](https://www.npmjs.com/package/@edge-csrf/nextjs) package which supports Next.js.

[^1]: https://nextjs.org/docs/app/building-your-application/routing/middleware#runtime
[^2]: https://nextjs.org/docs/app/api-reference/edge#unsupported-apis

**Reply by rwx222** [reply]
Thank you so much @Juneezee 🙏. 
I hadn't noticed that part of the documentation. Your response is very clear and concise. 
Best regards and blessings, I really appreciate it ❤️

**Reply by Juneezee** [reply]
@andresdotsh  You are absolutely welcome :heart:. You did great on providing a clear and concise description, along with those helpful screenshots, were fantastic. They made troubleshooting the issue much more easier!
